### ISSUE TITLE
v0.1完成後のlisjong-engineの完成像とgame-engine roadmapを文書化する

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong-engine#27]
## 背景

`lisjong-engine` は、lisjong ecosystemにおける日本リーチ麻雀のrule / state / game executionを担当する独立したgame engineである。

これまでの開発により、旧移行ロードマップ上のv0.1到達条件である deterministic minimal hanchan driver まで完了し、概ね次のfoundationが成立した。

```text
RuleSet / domain / scoring / settlement
        ↓
deterministic Round / Match execution
        ↓
SeatObservation + public action descriptors
        ↓
external seat selectors
        ↓
deterministic hanchan driver
        ↓
CompletedMatch
```

したがって今後の主題は「まず半荘を完走できるようにすること」ではなく、**v0.1相当の基礎完成を保ちながら、correctness、validation、recordability、interoperability、performance、consumer readinessを段階的に成熟させること**へ移る。

今後lisjong ecosystemでは、Policy / AI強化、reproducible evaluation、対局記録の保存・分析、replay / viewer、CLI / GUI、human vs AI、external environment / OSSとのinteroperability等への発展が想定される。

一方、将来必要そうなconsumerを想定してengine coreへ巨大な共通schemaやexternal protocol固有modelを先行導入することは避ける。

本Issueではcode / runtime behaviorを変更せず、**post-v0.1の `lisjong-engine` 完成像と長期roadmapを文書化する。**

## 目的

`lisjong-engine` の長期的な完成像を、

> 日本リーチ麻雀を正確かつdeterministicに実行し、stableなengine contractを通じてAI・評価・記録・UI・external integration等のconsumerから独立して利用できるgame engine

として整理する。

roadmapは厳密な直列Phaseではなく、次の継続Trackとして扱う。

```text
Current v0.1 foundation
        │
        ├─ Core Execution Hardening
        ├─ Stable Contracts
        ├─ Validation
        ├─ Recordability Support
        ├─ Consumer / Interoperability Readiness
        ├─ Performance
        └─ RuleSet Evolution
```

## 1. `docs/roadmap.md` を追加する

新たに `docs/roadmap.md` を追加し、post-v0.1 roadmapを正本化する。

文書責務は次のように分離する。

```text
README.md
    repositoryの現在のidentity / 利用入口

docs/architecture.md
    現在実装済みengine contractの正本

docs/roadmap.md
    post-v0.1の長期完成像 / 発展方向
```

READMEやarchitectureへ未実装roadmapを重複記載しない。

## 2. Current foundationを現在地点として整理する

v0.1 foundationをfuture featureとして書かず、今後の発展の出発点として整理する。

少なくとも、

- RuleSet / domain model
- scoring / settlement
- deterministic Round / Match progression
- `CompletedMatch`
- seat-specific `SeatObservation`
- public action descriptor
- external selector boundary
- deterministic minimal hanchan driver

が接続済みであることを前提とする。

具体APIや現行contractは `docs/architecture.md` を正本とし、roadmapへ複製しない。

## 3. Core Execution Hardening / Stable Contracts

Core Executionは新規にゼロから構築するTrackではなく、現在のrule-correct / deterministic executionを継続的にhardeningする。

対象はrule / state semantics、fail-closed behavior、transactional transition、error diagnosability等とする。

Stable Contractsは、concrete consumer / validation / interoperability use caseから必要性が確認された最小単位で発展させる。

```text
concrete use case
      ↓
missing requirement
      ↓
minimum engine contract
```

`lisjong`、`lisjong-arena`、viewer、CLI / GUI、MJAI等の固有型をengine core modelへそのまま逆流させない。

## 4. Validation Track

Validationをpost-v0.1の主要な継続Trackとして扱う。

```text
Validation
├─ example / regression tests
├─ invariants / property-oriented validation
├─ differential validation
└─ external conformance cases
```

候補にはtile / score conservation、legal-action consistency、turn ownership、meld / terminal / settlement consistency、deterministic progression等がある。

具体的なproperty-based testing frameworkは本Issueでは決定しない。

Differential validationでは意味的に独立したimplementationをreferenceとして利用できるが、

```text
implementation agreement != correctness proof
```

とする。同一backendのthin wrapperを独立referenceとして数えず、差異はrule / configuration / semantic mismatchを含めて調査する。majority voteをoracleにしない。

## 5. Recordability Support Track

engineが所有するのは「recordそのもの」ではなく、**recordを構成できるobjective execution factsを安全に提供できる能力**とする。

現在すでに存在するevent / result / settlement / history / random provenance等を出発点とし、

```text
existing engine facts
        ↓
concrete record / replay / analysis PoC
        ↓
missing informationを確認
        ↓
minimum record-supporting contract
```

という順序で発展させる。

巨大なcanonical `GameEvent` / `GameRecord` schema、JSON / JSONL、database、viewer、dataset pipeline等を先行設計しない。

### deterministic re-executionとhistorical record

次は別能力として扱う。

```text
same initial conditions + same actions
        ↓
deterministic re-execution
```

```text
saved historical evidence
        ↓
historical replay / analysis
```

**deterministic re-execution is not a substitute for historical evidence.**

### player-facing informationとprivileged ground truth

```text
SeatObservation
    !=
purpose-specific privileged immutable projection
    !=
engine internal mutable state
```

研究・validation用途でhidden hands、wall / dead-wall、random provenance等が必要になっても、`SeatObservation`へ混ぜず、mutable `RoundState` / `MatchState` をそのまま外部公開することも前提にしない。

AI固有のHandBelief、danger / value、Policy reasoning等は `lisjong` の責務とする。Arena固有のevaluation metrics / provenanceもgame recordそのものとは分離する。

## 6. Consumer / Interoperability Readiness

`lisjong-engine` 自身へ各consumer implementationを取り込むのではなく、stable boundaryから自然に利用できる状態を目指す。

MJAI、RiichiEnv等はengine core contractではなく、external adapter / PoC候補として扱う。

```text
lisjong-engine
      ↓
stable engine boundary
      ↓
external adapter / consumer PoC
      ↓
external ecosystem
```

PoCでinformation loss、RuleSet / semantic differences、replay fidelity等を確認し、engine側の不足が実証された場合だけminimum contract追加を検討する。

特定OSS / protocolをmandatory dependencyやcanonical representationとして固定しない。

recorder / viewer / telemetry / adapter等のfailureによってengine stateへpartial mutationやcorruptionを生じさせない。ただしconsumer failure時にexecutionを継続するかabortするかは、具体的integration contractで決定する。

## 7. Performance / RuleSet Evolution

Performanceは、

```text
Correctness
    ↓
Baseline measurement
    ↓
Actual bottleneck identification
    ↓
Targeted optimization
```

の順序を維持する。

actions / rounds / games per second、legal-action generation、state transition、scoring、observation construction、将来のrecordability overhead等を測定候補とする。

Rust化、native backend、backend replacement等を実測前にroadmapへ固定しない。

RuleSet Evolutionもvariation追加自体を目的とせず、external platform、Arena、Human Play等のconcrete ecosystem requirementから必要性が確認されたものを優先する。既存mechanicsを未実装featureとして再掲しない。

## 8. Now / Next / Later

roadmapは厳密な直列Phaseではない。

### Now

- current contract / docsの整合
- correctness hardening
- regression / invariant validation強化
- external validation候補の検討
- performance baseline measurement
- concrete consumer接続から不足boundaryを発見

### Next

- record / replay / analysis consumer PoC
- external adapter / interoperability PoC
- privileged ground-truth evaluation use case
- concrete consumer integration PoC（必要に応じて `lisjong` / `lisjong-arena` を含む）
- PoCから判明したminimum contract追加

これらすべてを同時に実施することや、次Issueの固定順序は要求しない。

### Later

- targeted performance optimization
- mature record / replay ecosystem support
- broader consumer readiness
- use-case driven RuleSet expansion
- external ecosystem integration拡張

## 9. Human Play readiness

Human Playそのものはengine責務としない。

現在の `SeatObservation + public action options + external selector` boundaryが、将来CLI / GUI等からhuman inputへ接続可能であることをConsumer Readinessの一部として考慮する。

UI、input handling、network session、authentication等はengine責務としない。

## 10. Project-wide正本との境界

以下のproject-wide原則は `lisjong-project` を正本とし、本roadmapでは再定義しない。

- repository間responsibility
- OSS / external ecosystemの一般的活用方針
- correctness → validation → measurement → optimization
- canonical event schemaをconcrete requirementsより先に設計しない原則
- Visualization / Analysis Track全体
- Human Play全体のecosystem architecture
- Learning Policy
- project-wide evaluation hierarchy

`lisjong-engine` roadmapは、それらをgame-engine責務として具体化することに集中する。

## README更新

READMEに残るv0.1以前の初期移行phaseを前提としたstale記述を最小限更新する。

READMEではIssue番号単位の進捗を列挙せず、staleになりにくい形で、

- engineの現在のidentity
- v0.1相当のfoundationが成立していること
- 現在contractは `docs/architecture.md`
- 長期方向は `docs/roadmap.md`
- 現在の進捗はGitHub Issues / PRs

への導線を整理する。

## 対象ファイル

主な変更対象:

```text
docs/roadmap.md   # new
README.md         # minimal update
```

`docs/architecture.md` は現在contractの正本として維持し、明確なstale / contradictionが見つかった場合だけ最小限同期する。

## 完了条件

- [ ] `docs/roadmap.md` が追加されている
- [ ] G2完了を含むv0.1 foundationが現在地点として整理されている
- [ ] Core Executionが継続的hardeningとして位置付けられている
- [ ] Stable Contractsがconcrete requirement drivenで発展する方針になっている
- [ ] Validationが継続Trackとして整理され、agreementをproof / majority oracleとして扱っていない
- [ ] RecordabilityがGameRecord schema所有ではなくobjective execution facts提供能力として整理されている
- [ ] deterministic re-executionとhistorical record replayが分離されている
- [ ] `SeatObservation`、privileged ground truth、mutable internal stateの境界が分離されている
- [ ] AI analysis / Arena evaluation provenanceとgame-engine factsが責務分離されている
- [ ] MJAI / RiichiEnv等をexternal adapter / PoC候補として扱い、core dependencyへ固定していない
- [ ] consumer failure isolationがstate corruption防止として整理されている
- [ ] performance measurementとoptimizationが区別され、native化を先行固定していない
- [ ] RuleSet Evolutionがconcrete use-case drivenになっている
- [ ] Now / Next / Laterがpost-v0.1の現在地点を反映している
- [ ] READMEの初期移行phaseに関するstale記述が更新されている
- [ ] project-wide原則との正本境界が明記されている
- [ ] code / runtime behaviorを変更していない

## 対象外

本Issueでは以下を実装・設計しない。

- game engine implementation / public API変更
- RuleSet追加
- GameRecord / canonical GameEvent schema新設
- privileged ground-truth schema設計
- persistence / JSON / JSONL format設計
- event bus / observer API設計
- MJAI / RiichiEnv / `lisjong` / `lisjong-arena` integration実装
- differential test / property-based testing framework導入
- training dataset生成
- viewer / replay UI / CLI / GUI / Human Play実装
- network / session実装
- performance optimization / Rust / native backend導入
- external OSS dependency追加

これらはroadmap文書化後、concrete needが確認されたものから個別Issueとして扱う。

## 補足

v0.1到達により、

```text
麻雀を決定的に最後まで動かせるか
```

という基礎課題は一区切りとなった。

今後は、

```text
correct execution
      ↓
validation / hardening
      ↓
concrete consumer usage
      ↓
missing requirements発見
      ↓
minimum contract extension
      ↓
measurement / further validation
```

というloopでengineを成熟させる。

`lisjong-engine` は将来必要そうなものをすべて先回りして実装するrepositoryではなく、**game-engine固有のrule correctnessとstable contractを所有し、具体的consumer requirementsに応じて必要最小限の能力を育てる基盤**として発展させる。
